src/languages/gml.js

```diff
@@ -2994,135 +937,572 @@ export default function(hljs) {
+  /**
+   * Regex for a dot separating some LHS and RHS expression with optional whitespace (as this is
+   * supported in the engine.)
+   */
+  const DOT_ACCESSOR_REG = /\b\.\b/;
+
+  /**
+   * Expressions, which form part of a valid statement.
+   * @type {Object[]}
+   */
+  const EXPRESSION = [];
+
+  /**
+   * A single-line comment.
+   */
+  const COMMENT_LINE = hljs.COMMENT('//', /\$|\n/);
+
+  const JSDOC_TYPE_PARAM_REG = /{ *?[a-zA-Z_][a-zA-Z0-9_\.<>\|]*? *?} *?/;
+
+  const JSDOC_ANNOTATIONS = [
+    {
+      match: /@((desc(ription)?)|ignore|pure|deprecated|(func(tion)?))\b/,
+      className: "keyword"
+    },
+    {
+      match: [
+        /@(self|context) *?/,
+        JSDOC_TYPE_PARAM_REG
+      ],
+      className: {
+        1: "keyword",
+        2: "literal",
+      }
+    },
+    {
+      match: [
+        /(@return)s? *?/,
+        JSDOC_TYPE_PARAM_REG
+      ],
+      className: {
+        1: "keyword",
+        2: "literal",
+      }
+    },
+    {
+      match: [
+        /@((param(eter)?)|arg(ument)?) *?/,
+        JSDOC_TYPE_PARAM_REG,
+        /\[?/,
+        VALID_IDENTIFIER_REG,
+        /(( *=[^\n]+?)?\])? *?/
+      ],
+      className: {
+        1: "keyword",
+        2: "literal",
+        3: "variable",
+        4: "variable",
+        5: "variable",
+      }
+    }
   ];
+
+  /**
+   * A comment that documents a function using the same style as JSDoc for JavaScript.
+   */
+  const COMMENT_JSDOC = {
+    begin: /\/\/\//,
+    end: /$|\n/,
+    className: "comment",
+    contains: JSDOC_ANNOTATIONS
+  };
+
+  /**
+   * A comment that documents a function using the same style as JSDoc for JavaScript.
+   */
+  const COMMENT_BLOCK_JSDOC = {
+    begin: /\/\*\*/,
+    end: /\*\//,
+    className: "comment",
+    contains: JSDOC_ANNOTATIONS
+  };
+
+  /**
+   * Modes for the types of comments supported in GML.
+   */
+  const COMMENT = {
+    variants: [
+      COMMENT_JSDOC,
+      COMMENT_BLOCK_JSDOC,
+      COMMENT_LINE,
+      hljs.C_BLOCK_COMMENT_MODE,
+    ]
+  };
+
+  /**
+   * A template string substitution. `contains` is filled in after `EXPRESSION` is defined due to
+   * nesting.
+   */
+  const STRING_SUBSTITUTION = {
+    begin: /{/,
+    end: /}/,
+    beginScope: "literal",
+    endScope: "literal",
+    contains: EXPRESSION
+  };
+
+  /**
+   * A template string substitution for use with the older `string()` optional args with `"{0}"`,
+   * etc.
+   */
+  const STRING_NUMERICAL_SUBSTITUTION = {
+    match: /{[0-9]+}/,
+    scope: "literal"
+  };
+
+  /**
+   * An escape sequence in a string.
+   */
+  const STRING_ESCAPE = {
+    scope: "literal",
+    variants: [
+      { match: /\\u[a-fA-F0-9]{1,6}/ },
+      { match: /\\[^\n]/ }
+    ]
+  };
+
+  /**
+   * Various types of strings supported in the engine.
+   */
+  const STRING = {
+    variants: [
+      {
+        begin: /\$"/,
+        end: "\"",
+        beginScope: "string",
+        endScope: "string",
```

Comment:
I was having a lot of issues getting template literals to behave as I wanted. The issue was that if I simply set `scope: "string"`, then the entire thing would be wrapped with that class, which would then make anything within templates, e.g.,`$"{someVariable}"` appear string-coloured if not claimed by another expression type (e.g., a number would still look like a number, since it would claim it.)

I might have a look at how some other languages approached this issue as I wasn't happy with this solution regardless, but I recall going down a rabbit hole of greater and greater complexity language-knowledge by the parser to fix this - there's probably an easy solution I missed though :P

Reply:
I've come back to this again and came up with a middleground that I think is a lot more reasonable - this has been kept, but I've fixed the issue where string characters were being matched one-by-one which was making a ridiculous number of individual `<span>`s for each character. This has been fixed, but the `string` scope is still not applied to the entire span, so that the inner templates are not interfered with by `string` in themes.

Reply:
Maybe we'd allow some liberty with the GML theme... not sure... but typically nested items like:

```
string -> subst -> number
```

The subst color would override string, and the number color would override subst.  If a theme doesn't provide a style for `subst` that theme is broken and should be fixed.  Background can be a real problem, but not a lot of themes do crazy things with backgrounds - is GML one of them?

Still not entire I'm completely following your issue.

Reply:
The root of the issue is that identifiers for things like local variables do not have any matching parse rule, and are unhighlighted. I didn't add any "catch all" rule at the bottom for this, though I guess I could throw in such a rule for anything that's a valid identifier with a relevance of 0. As this rule does not currently exist, here's what happens if I let the scopes be nested:
![image](https://github.com/user-attachments/assets/a7ceac98-3153-400e-9fb6-0e90a5051c2f)

...the result being that these variables become string-coloured.
However, using the setup that's currently in, the issue is fine:
![image](https://github.com/user-attachments/assets/c30e72c7-30dc-4738-84cf-9251cabf1233)

Another thing going for this to me was that it avoids the issue where a nested scope would inherit things you don't want it to - e.g., if you had a theme which made strings bold, you don't really want the contents of their templates to also become bold with them.

Reply:
Ok, I looked closely at your mode for string templates now - yeah this is not the preferred way - matching string like chunks inside strings. :).  Neat trick though.

Isn't this same effect achievable thru pure CSS with just a ~ `.string .subst` rule that sets the styling (color, bg, bold/italic) back to whatever the theme considers "standard"?

This is the kind of thing I"d prefer to solve at the engine/CSS level rather than encouraging every individual grammar to start adding "hacks" like this to accomplish stylistic goals.

Reply:
Totally understand about avoiding these sorts of hacks.

As for the alternative, if I had `.subst` reset the colour scheme in CSS, it'd also cause the surrounding `{}` to become an un-highlighted colour. I've taken that idea though and modified it such that `.subst` resets colours, but the `{}` is given `char.escape` instead - would this be an acceptable solution?